### ISSUE TITLE
Eager load only once - Performance

### ISSUE THREAD

**Issue by garygreen** [laravel/internals#450]
If you do:

```php
DB::enableQueryLog();

$user = App\User\User::find(1);

$user->load(['address']);
$user->load(['address']);
$user->load(['address']);

dd(DB::getQueryLog());
```

It'll eager load the address information 3 times. This seems a bit strange since Laravel could detect if the relation is already loaded and if so, don't attempt to re-eager load.

This would be particulary useful in api transformers where you can simplify your transformer like so:

```php
class UserTransformer {

    public function transformUsers($users)
    {
        $this->eagerLoadRelations($users);
        $result = [];
        
        foreach ($users as $user) {
            $result[] = $this->transformUser($user);
        }

        return $result;
    }

    public function transformUser(User $user)
    {
        $this->eagerLoadRelations($user); // Currently, this will re-load the relations.

        return $user->toArray();
    }

    protected eagerLoadRelations($obj)
    {
        $obj->load(['address']);
    }

}
```

The above example is very simplistic and you could pass a parameter to `transformUser` to tell it not to eager load, or do lots of other isset-type checks but sometimes it's not always easy to reason what relations have already been accessed/loaded in various parts of your app, so it's either eager loading something twice, or wrapping your stuff in lots of isset checks, which is problematic for nested relations.

**Comment by franzliedke** [comment]
Hmm... reloading the relation explicitly (i.e. getting up-to-date data from the database) would still be a valid use case - how would that still be possible?

**Comment by sisve** [comment]
There's also issues with filtered relations, where you've only eager-loaded some models and now want to change it. Or perhaps just the order of them.

**Comment by Bakke** [comment]
You could solve this by doing:
```php
$user = App\User\User::find(1);
if (!isset($user->relations['address'])) {
    $user->load(['address']);
}
```

**Comment by garygreen** [comment]
@Bakke of course, but when you want to load lots of relations it becomes a bit of a headache - particularly nested relations.

**Comment by Bakke** [comment]
Maybe a good solution would be to create a boolean property on the model for reloading relationships with `load()`, and/or creating a second `eagerLoadRelation()` function on `Illuminate\Database\Eloquent\Builder`? This way relationships could be reloaded as default.

If a relationship is not reloaded we would also have to check if nested relationships further down is loaded. Example:
```
$user = App\User\User::find(1);

// Turn reload off

$user->load('posts');
$user->load('posts.comments');
```

Comments would never be fetched because this depends on posts being "hydrated".

**Comment by Bakke** [comment]
How about something like this?
https://github.com/Bakke/framework/commit/2eb1693148dc6b21ed7cfa8cead8b12ba07380fd

However, this does not account for new constraints or new nested relationships.

Example:
```
$tags = Tag::all();
$tags->lazyLoad('posts');
$tags->lazyLoad('posts.comments');
```
Comments would never be loaded. Adding a deep check for `relationLoaded()` could fix this problem.

**Comment by deleugpn** [comment]
Why would you call load again if you're not interested in loading a fresh set of data from your dataset?

**Comment by Bakke** [comment]
See first post.

If you pass objects between functions you might want to make sure relationships are loaded, but not reload them.

**Comment by JayBizzle** [comment]
Whilst we are talking about relationships, it would be nice if this could get fixed, see laravel/framework#14090
